From 22720e12cb02e5f45b18e8a44a93e75b7e66b296 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Tue, 13 May 2014 08:54:16 -0400
Subject: [PATCH 025/102] listfs: If LDM not available, don't inhibit partition
 detection (RHBZ#1079182).

If a disk has type 0x42 partition (which would indicate LDM), but LDM
is not available then try parsing the partition anyway.  It might be
parseable as plain old NTFS.
(cherry picked from commit 47ec61e9af19f8a750750ad498adc10fa7ae0521)
---
 src/listfs.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/src/listfs.c b/src/listfs.c
index b87a021..d53b3828 100644
--- a/src/listfs.c
+++ b/src/listfs.c
@@ -49,6 +49,9 @@ guestfs__list_filesystems (guestfs_h *g)
   char **ret;
   size_t ret_size = 0;
 
+  int has_lvm2 = guestfs___feature_available (g, "lvm2");
+  int has_ldm = guestfs___feature_available (g, "ldm");
+
   CLEANUP_FREE_STRING_LIST char **devices = NULL;
   CLEANUP_FREE_STRING_LIST char **partitions = NULL;
   CLEANUP_FREE_STRING_LIST char **mds = NULL;
@@ -88,7 +91,7 @@ guestfs__list_filesystems (guestfs_h *g)
 
   /* Use vfs-type to check for filesystems on partitions. */
   for (i = 0; partitions[i] != NULL; ++i) {
-    if (! is_mbr_partition_type_42 (g, partitions[i]))
+    if (has_ldm == 0 || ! is_mbr_partition_type_42 (g, partitions[i]))
       check_with_vfs_type (g, partitions[i], &ret, &ret_size);
   }
 
@@ -96,7 +99,7 @@ guestfs__list_filesystems (guestfs_h *g)
   for (i = 0; mds[i] != NULL; ++i)
     check_with_vfs_type (g, mds[i], &ret, &ret_size);
 
-  if (guestfs___feature_available (g, "lvm2")) {
+  if (has_lvm2 > 0) {
     /* Use vfs-type to check for filesystems on LVs. */
     lvs = guestfs_lvs (g);
     if (lvs == NULL) goto error;
@@ -105,7 +108,7 @@ guestfs__list_filesystems (guestfs_h *g)
       check_with_vfs_type (g, lvs[i], &ret, &ret_size);
   }
 
-  if (guestfs___feature_available (g, "ldm")) {
+  if (has_ldm > 0) {
     /* Use vfs-type to check for filesystems on Windows dynamic disks. */
     ldmvols = guestfs_list_ldm_volumes (g);
     if (ldmvols == NULL) goto error;
-- 
1.9.3

